test/net_http_test.rb

```diff
@@ -0,0 +1,318 @@
+require 'test_helper'
+require 'semian/net_http'
+require 'thin'
+
+class TestNetHTTP < MiniTest::Unit::TestCase
+  class RackServer
+    def self.call(env)
+      response_code = env['REQUEST_URI'].delete("/")
+      response_code = '200' if response_code == ""
+      [response_code, {'Content-Type' => 'text/html'}, ['Success']]
+    end
+  end
+
+  PORT = 31_050
+  TOXIC_PORT = PORT + 1
+
+  def test_semian_identifier
+    with_server do
+      Net::HTTP.start("localhost", TOXIC_PORT) do |http|
+        assert_equal "http_localhost_31051", http.semian_identifier
+      end
+    end
+  end
+
+  def test_trigger_open
+    with_server do
+      open_circuit!
+      uri = URI('http://localhost:31051/200')
+      assert_raises Net::CircuitOpenError do
+        Net::HTTP.get(uri)
+      end
+    end
+  end
+
+  def test_trigger_close_after_open
+    with_server do
+      open_circuit!
+      close_circuit!
+    end
+  end
+
+  def test_bulkheads_tickets_are_working
+    with_server do
+      threads = []
+      ticket_count = Net::HTTP.new("localhost", TOXIC_PORT).raw_semian_options[:tickets]
+
+      @proxy.downstream(:latency, latency: 500).apply do
+        uri = URI('http://localhost:31051/200')
+        (ticket_count + 1).times do
+          threads << Thread.new do
+            Net::HTTP.get(uri)
+          end
+        end
+        assert_raises Net::ResourceBusyError do
```

Comment:
Similarly you could have a situation where none of them raise.. that's up to the scheduler. This test is fundamentally racy